### ISSUE TITLE
Strides computed via autograd.grad vs those computed by backward() can be different

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#49256]
## 🐛 Bug

When computing the gradient via autograd.grad, the strides of the returned gradient can be different from if we had instead used backward() to compute the gradient. I don't know if this is intended or not.

## To Reproduce
```
x = torch.randn(3, 3, requires_grad=True)
y = x.clone()
gy = torch.randn(3, 3).t()
res1, = torch.autograd.grad(y, x, gy)

x = torch.randn(3, 3, requires_grad=True)
y = x.clone()
gy = torch.randn(3, 3).t()
y.backward(gy)
res2 = x.grad

print(res1.stride())
(1, 3)
print(res2.stride())
(3, 1)
```

## Expected behavior

I would expect these two to return gradients with the same strides. Furthermore, the [Gradient Layout Contract note](https://github.com/pytorch/pytorch/blob/db5e5b439c454d657cfa8f08a096cf68e203f2a8/torch/csrc/autograd/functions/accumulate_grad.h#L42-L46) seems to imply that we prefer to return gradients that have the same strides as our inputs.

## Environment

master

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

**Comment by albanD** [comment]
I would argue that this is expected.
The main reason for the layout contract in backward and all the logic in AccumulateGrad in general is because this attribute is here to stay as the `.grad` field. And so it should have some nice properties.
For `autograd.grad` the user can do what they want with the result and so it is not as important for it to have a specific layout. Also enforcing such a layout here might be an expensive step for a Tensor that is just read once.

**Comment by ezyang** [comment]
cc @ngimel @VitalyFedyunin @glaringlee

**Comment by ngimel** [comment]
I agree with @albanD's take. Gradient Layout Contract is there to be nice to optimizers, but `autograd.grad` has no such constraints. Usually the gradient layout automatically follows the original tensor layout, but in case it doesn't, I don't see a compelling reason to force it.

**Comment by albanD** [comment]
Closing as expected behavior for now.
Feel free to re-open if it causes issues in user code.
